File: src/test/scala/folding.scala
def foldRight[A,B](as: List[A], b: B)(f: (A, B) => B): B =
  as match {
    case Nil => b
    case x :: xs => f(x, foldRight(xs, b)(f))
  }

foldRight(List(1, 2, 3))


def foldLeft[A,B](as: List[A], b: B)(f: (B, A) => B): B =
as match {
  case Nil   => b
  case h :: t => foldLeft(t, f(b, h))(f)
}

File: src/main/kotlin/com/harmellaw/kotlin/Folding.kt
package com.harmellaw.kotlin

tailrec fun <A, B> foldLeft(l: List<A>,
                    seed: B,
                    func: (b: B, a: A) -> B) : B {

    return when {
        l.isEmpty() -> seed
        else        -> foldLeft(l.tail, func(seed, l.head), func)
//        else -> throw IllegalArgumentException("The list was not empty")
    }
}

fun <A, B> foldLeftWithoutTailRecursion(l: List<A>,
                            seed: B,
                            func: (b: B, a: A) -> B) : B {

    return when {
        l.isEmpty() -> seed
        else        -> foldLeftWithoutTailRecursion(l.tail, func(seed, l.head), func)
    }
}

fun <A, B> foldRight(l: List<A>,
                     seed: B,
                     func: (b: B, a: A) -> B) : B {

    return foldLeft(l.asReversed(), seed, func)
}

fun <A, B> foldRightOverflows(l: List<A>,
                              seed: B,
                              func: (a: A, b: B) -> B) : B {

    return when {
        l.isEmpty() -> seed
        else -> func(l.head, foldRightOverflows(l.tail, seed, func))
    }
}